modules/catalog-compaction/src/integrationTest/java/org/apache/ignite/internal/catalog/compaction/ItCatalogCompactionTest.java

```diff
@@ -0,0 +1,226 @@
+import org.apache.ignite.internal.table.distributed.raft.PartitionListener;
+import org.apache.ignite.internal.testframework.IgniteTestUtils;
+import org.apache.ignite.internal.tx.InternalTransaction;
+import org.apache.ignite.network.ClusterNode;
+import org.apache.ignite.raft.jraft.RaftGroupService;
+import org.apache.ignite.tx.TransactionOptions;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Integration tests to verify catalog compaction.
+ */
+class ItCatalogCompactionTest extends ClusterPerClassIntegrationTest {
+    private static final int CLUSTER_SIZE = 3;
+
+    @Override
+    protected int initialNodes() {
+        return CLUSTER_SIZE;
+    }
+
+    @Test
+    void testRaftGroupsUpdate() throws InterruptedException {
+        IgniteImpl ignite = CLUSTER.aliveNode();
+        CatalogManagerImpl catalogManager = ((CatalogManagerImpl) CLUSTER.aliveNode().catalogManager());
+
+        sql(format("create zone if not exists test with partitions=16, replicas={}, storage_profiles='default'",
+                initialNodes()));
+        sql("alter zone test set default");
+
+        // Latest active catalog contains all required tables.
+        {
+            sql("create table a(a int primary key)");
+
+            Catalog minRequiredCatalog = catalogManager.catalog(catalogManager.latestCatalogVersion());
+            assertNotNull(minRequiredCatalog);
+
+            sql("create table b(a int primary key)");
+
+            HybridTimestamp expectedTime = HybridTimestamp.hybridTimestamp(minRequiredCatalog.time());
+
+            CompletableFuture<Void> fut = ignite.catalogCompactionRunner()
+                    .propagateTimeToReplicas(expectedTime.longValue());
+
+            assertThat(fut, willCompleteSuccessfully());
+
+            ensureTimestampStoredInAllReplicas(expectedTime, 2);
+        }
+
+        // Latest active catalog does not contain all required tables.
+        // Replicas of dropped tables must also be updated.
+        long requiredTime = CLUSTER.aliveNode().clockService().nowLong();
+
+        {
+            sql("drop table a");
+            sql("drop table b");
+
+            HybridTimestamp expectedTime = HybridTimestamp.hybridTimestamp(requiredTime);
+
+            CompletableFuture<Void> fut = ignite.catalogCompactionRunner()
+                    .propagateTimeToReplicas(expectedTime.longValue());
+
+            assertThat(fut, willCompleteSuccessfully());
+
+            ensureTimestampStoredInAllReplicas(expectedTime, 2);
+        }
+
+        // Update to lower timestamp should not succeed.
+        {
+            HybridTimestamp expectedTime = HybridTimestamp.hybridTimestamp(requiredTime - 1);
+
+            CompletableFuture<Void> fut = ignite.catalogCompactionRunner()
+                    .propagateTimeToReplicas(expectedTime.longValue());
+
+            assertThat(fut, willCompleteSuccessfully());
+
+            ensureTimestampStoredInAllReplicas(HybridTimestamp.hybridTimestamp(requiredTime), 2);
+        }
```

Comment:
1. Infinitely dead assignment case is not covered. I mean the one when `Assignment assignment = tokenizedAssignments.nodes().iterator().next();` is not present in topology and never will be.
2. Table adjustments while invokeOnReplicas recursive calls is not covered.
3. Failover cases are not covered. E.g. temporary assignment death.

Reply:
> 1. Infinitely dead assignment case is not covered. I mean the one when Assignment assignment = tokenizedAssignments.nodes().iterator().next(); is not present in topology and never will be.

I added some tests on mocks for check this (`CatalogCompactionRunnerSelfTest`):
* minTxTimePropagationSucceedWhenSomeAssignmentIsMissing
* minTxTimePropagation
* minTxTimePropagationAbortedIfNoAssignmentsPresentInTopology

> 2. Table adjustments while invokeOnReplicas recursive calls is not covered.

This test case needs more clarification, I don't quite understand what this test might look like. :thinking: 

> 3.     Failover cases are not covered. E.g. temporary assignment death.

Currently we don't have any failover :sunglasses:  We just aborting the process (see `minTxTimePropagationAbortedIfNoAssignmentsPresentInTopology`).